components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3692,5 +3249,628 @@ subroutine get_time_space_phys_variables( &
+
+         !--- for top level only (since flux is 0 above)
+         k = k_qxtop
+         !-- compute flux divergence
+         fluxdiv_qit = -flux_qit(k)*inv_dzq(k)
+         fluxdiv_qir = -flux_qir(k)*inv_dzq(k)
+         fluxdiv_bir = -flux_bir(k)*inv_dzq(k)
+         fluxdiv_nit = -flux_nit(k)*inv_dzq(k)
+         !-- update prognostic variables
+         qitot(k) = qitot(k) + fluxdiv_qit*dt_sub*inv_rho(k)
+         qirim(k) = qirim(k) + fluxdiv_qir*dt_sub*inv_rho(k)
+         birim(k) = birim(k) + fluxdiv_bir*dt_sub*inv_rho(k)
+         nitot(k) = nitot(k) + fluxdiv_nit*dt_sub*inv_rho(k)
+         !zitot(i,k) = zitot(i,k) + fluxdiv_nit*dt_sub*inv_rho(i,k)
+
+
+         do k = k_qxtop-kdir,k_temp,-kdir
+            !-- compute flux divergence
+            fluxdiv_qit = (flux_qit(k+kdir) - flux_qit(k))*inv_dzq(k)
+            fluxdiv_qir = (flux_qir(k+kdir) - flux_qir(k))*inv_dzq(k)
+            fluxdiv_bir = (flux_bir(k+kdir) - flux_bir(k))*inv_dzq(k)
+            fluxdiv_nit = (flux_nit(k+kdir) - flux_nit(k))*inv_dzq(k)
+            !-- update prognostic variables
+            qitot(k) = qitot(k) + fluxdiv_qit*dt_sub*inv_rho(k)
+            qirim(k) = qirim(k) + fluxdiv_qir*dt_sub*inv_rho(k)
+            birim(k) = birim(k) + fluxdiv_bir*dt_sub*inv_rho(k)
+            nitot(k) = nitot(k) + fluxdiv_nit*dt_sub*inv_rho(k)
+            !zitot(i,k) = zitot(i,k) + fluxdiv_nit*dt_sub*inv_rho(i,k)
+         enddo
+
+         dt_left = dt_left - dt_sub  !update time remaining for sedimentation
+         if (k_qxbot.ne.kbot) k_qxbot = k_qxbot - kdir
+         !or, optimzed: k_qxbot = k_qxbot +(k_qxbot.eq.kbot)*kdir
+
+      enddo substep_sedi_i
+
+      prt_sol = prt_sol + prt_accum*inv_rhow*odt
+
+   endif qi_present
+
+   qi_tend(:) = ( qitot(:) - qi_tend(:) ) * odt ! Ice sedimentation tendency, measure
+   ni_tend(:) = ( nitot(:) - ni_tend(:) ) * odt ! Ice # sedimentation tendency, measure
+
+end subroutine
+
+subroutine homogeneous_freezing(kts,kte,t,exner,xlf,  & 
+   qc,nc,qr,nr,qitot,nitot,qirim,birim,th) 
+  
+   !.......................................
+   ! homogeneous freezing of cloud and rain
+
+   implicit none 
+   integer, intent(in) :: kts, kte
+   real(rtype), intent(in), dimension(kts:kte) :: t 
+   real(rtype), intent(in), dimension(kts:kte) :: exner
+   real(rtype), intent(in), dimension(kts:kte) :: xlf
+
+   real(rtype), intent(inout), dimension(kts:kte) :: qc
+   real(rtype), intent(inout), dimension(kts:kte) :: nc
+   real(rtype), intent(inout), dimension(kts:kte) :: qr
+   real(rtype), intent(inout), dimension(kts:kte) :: nr 
+
+   real(rtype), intent(inout), dimension(kts:kte) :: qitot
+   real(rtype), intent(inout), dimension(kts:kte) :: nitot
+   real(rtype), intent(inout), dimension(kts:kte) :: qirim 
+   real(rtype), intent(inout), dimension(kts:kte) :: birim
+   real(rtype), intent(inout), dimension(kts:kte) :: th
+   real(rtype) :: Q_nuc
+   real(rtype) :: N_nuc 
+   integer :: k, ktop, kbot, kdir
+
+   ktop = kts
```

Comment:
Regardless of how we decide to treat `ktop` and `kbot` we should probably make them public variables in the module and declare them once.  Otherwise, if we have to change the hard-coded values we will have to now change them in a set of sub-routines which can lead to bugs if we miss just one of them.

Reply:
Thinking about this today and chatting with @pressel, I think the right strategy would be to keep `kbot`, `ktop`, `kts`, and `kte` as is but move the declaration of `ktop` and `kbot` to the `p3_main` subroutine and make it public.  This would avoid having to make a change to these variables in all of the subroutines if we ever had to in the future.

Reply:
kbot, ktop, and kdir are now passed to subroutines as need rather than set locally.